iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift

```diff
@@ -0,0 +1,304 @@
+
+        // Get all of the items and sort them by value. This combines all the sections into a flat list.
+        var sortedItemIndexPaths: [IndexPath] = []
+        self.forEachIndexPath { indexPath in
+            sortedItemIndexPaths.append(indexPath)
+        }
+        sortedItemIndexPaths.sort { indexPath1, indexPath2 in
+            let message1 = dataSource.getMessage(at: indexPath1)!
+            let message2 = dataSource.getMessage(at: indexPath2)!
+            return message1.createdAt < message2.createdAt
+        }
+
+        for (sortedItemsIndex, indexPath) in sortedItemIndexPaths.enumerated() {
+            let currentSection = indexPath.section
+            let currentItemIndex = indexPath.item
+
+            var startZ: CGFloat = 0
+            // Each item's z range starts after the end of the previous item's range within its section.
+            if let previousRangeInSection = self.zRangesDict[IndexPath(item: currentItemIndex - 1,
+                                                                      section: currentSection)] {
+
+                startZ = previousRangeInSection.upperBound + self.itemSize.height
+            }
+
+            var endZ = startZ
+            // Each item's z range ends before the beginning of the next item's range from within its section.
+            if sortedItemsIndex + 1 < sortedItemIndexPaths.count {
+                for nextIndex in (sortedItemsIndex+1)..<sortedItemIndexPaths.count {
+                    let nextIndexPath = sortedItemIndexPaths[nextIndex]
+
+                    if currentSection == nextIndexPath.section {
+                        endZ = CGFloat(nextIndex) * self.itemSize.height - self.itemSize.height
+                        break
+                    } else if nextIndex == sortedItemIndexPaths.count - 1 {
+                        endZ = CGFloat(nextIndex) * self.itemSize.height
+                        break
+                    }
+                }
+            }
+
+            self.zRangesDict[indexPath] = startZ..<endZ
+        }
+    }
+
+    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
+        // Return all items whose frames intersect with the given rect and aren't invisible.
+        let itemAttributes = self.cellLayoutAttributes.values.filter { attributes in
+            return attributes.alpha > 0 && rect.intersects(attributes.frame)
+        }
+
+        return itemAttributes
+    }
+
+    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
+        // If the attributes are cached already, just return those.
+        if let attributes = self.cellLayoutAttributes[indexPath]  {
+            return attributes
+        }
+
+        // All items in a section are positioned relative to its frontmost item.
+        guard let frontmostIndexPath = self.getFrontmostIndexPath(in: indexPath.section) else { return nil }
```

Comment:
We don't want any gaps in the message stack so item positions are always relative to the frontmost item.